src/Microsoft.VisualStudio.Editors/PropPages/ReferencePropPage.vb

```diff
@@ -1404,93 +1384,75 @@ Namespace Microsoft.VisualStudio.Editors.PropertyPages
         ''' Imported namespaces are currently added/removed one at a time
         '''    All removes are processed first, and then adds 
         '''</summary>
-        ''' <param name="NewImportList">the imported list being saved</param>
+        ''' <param name="newImportList">the imported list being saved</param>
         ''' <return>return true if any value was changed...</return>
         ''' <remarks>
         ''' CONSIDER: This is how the msvbprj code did it, and it may not work well
         '''         for other compilers (assuming this page is later shared)
         '''</remarks>
-        Private Function SaveImportedNamespaces(NewImportList As String()) As Boolean
-            Dim theVSProject As VSLangProj.VSProject
-            Dim _Imports As VSLangProj.Imports
-            Dim index, ListIndex As Integer
-            Dim _Namespace As String
+        Private Function SaveImportedNamespaces(newImportList As String()) As Boolean
             Dim valueUpdated As Boolean = False
-
-            theVSProject = CType(DTEProject.Object, VSLangProj.VSProject)
-
-            _Imports = theVSProject.Imports
+            Dim vsImports As VSLangProj.Imports = CType(DTEProject.Object, VSLangProj.VSProject).Imports
 
             Debug.Assert(Not _ignoreImportEvent, "why m_ignoreImportEvent = TRUE?")
             Try
                 _ignoreImportEvent = True
 
                 'For backward compatibility we remove all non-imported ones from the current Imports before adding any new ones
-                Dim CurrentImports As String() = GetCurrentImports()
-
-                index = CurrentImports.Length
-                For index = _Imports.Count To 1 Step -1
-                    _Namespace = CurrentImports(index - 1)
-                    ListIndex = LookupInStringArray(NewImportList, _Namespace)
-                    If ListIndex = -1 Then
-                        Debug.WriteLine("Removing reference: " & _Imports.Item(index))
-                        Try
-                            _Imports.Remove(index)
+                Dim currentImports = New List(Of String)(vsImports.Count)
```

Comment:
Look at the code in VisualBasicNamespaceImportsList for us for CPS projects, this is a race waiting to happen. 

Basically, we store the imports in List<string> and change it from _another_ thread (via dataflow) when you remove it here. In small testing you probably beat the race, but I'm sure this will be hit in the wild. I'd either change the code here to clone before modification - or fix VisualBasicNamespaceImportsList.

Reply:
It's worse than that - it appears to be impossible to actually iterate these in a safe manner: https://github.com/dotnet/project-system/issues/3101. At any time behind the scenes the imports could change.

Reply:
Note that this code enumerates backwards through the list, so as long as an unrelated dataflow change doesn't get processed, this code should be fine I think...

Reply:
If I'm reading VisualBasicNamespaceImportsList code correct, this line will cause the data flow to run:

```
vsImports.Remove(index)
```

Reply:
I suppose it would be easy to test this - put a 10 second sleep after that call and set a break point.

Reply:
Yes, I expect it would cause it to run, but as long as the order stays the same, that would remove the one at a higher index than we will look at for the rest of the method.

Reply:
I'll verify with a sleep.

Reply:
Verified with a sleep that this still works fine if this call comes in the middle.

Reply:
k, I'm okay with the change and we'll clean up the Imports service later.